From: Rafael J. Wysocki <rjw@suse.de>
Subject: x86_64 / PCI: Insert IO-APIC resources before unassigned resources
References: bnc#531260
Patch-upstream: Yes

Based on upstream commit 857fdc53a0a90c3ba7fcf5b1fb4c7a62ae03cf82
(x86/pci: insert ioapic resource before assigning unassigned
resources) from Yinghai Lu <yinghai@kernel.org>.

It turns out that:
(1) AMD-based systems have two HT chains.
(2) BIOS doesn't allocate resources for BAR 6 of devices under 8132
    etc.
(3) Multi-peer-root patch will try to split root resources to peer
    root resources according to PCI conf of NB
(4) PCI core assigns unassigned resources, but they overlap with BARs
    that are used by IO-APIC addr of io4 and 8132.

This happens because at that point the IO-APIC resources have not
been inserted yet.  The solution is to insert IO-APIC resources into
the tree a bit earlier.

Signed-off-by: Rafael J. Wysocki <rjw@suse.de>
Automatically created from "patches.arch/bug-531260-x86-pci-insert-ioapic-resource-before-assigning-unassigned-resources.patch" by xen-port-patches.py

--- sle11-2009-10-08.orig/arch/x86/kernel/io_apic_64-xen.c	2009-03-24 10:19:57.000000000 +0100
+++ sle11-2009-10-08/arch/x86/kernel/io_apic_64-xen.c	2009-10-08 12:18:42.000000000 +0200
@@ -2423,7 +2423,7 @@ void __init ioapic_init_mappings(void)
 	}
 }
 
-static int __init ioapic_insert_resources(void)
+void __init ioapic_insert_resources(void)
 {
 	int i;
 	struct resource *r = ioapic_resources;
@@ -2431,18 +2431,12 @@ static int __init ioapic_insert_resource
 	if (!r) {
 		printk(KERN_ERR
 		       "IO APIC resources could be not be allocated.\n");
-		return -1;
+		return;
 	}
 
 	for (i = 0; i < nr_ioapics; i++) {
 		insert_resource(&iomem_resource, r);
 		r++;
 	}
-
-	return 0;
 }
-
-/* Insert the IO APIC resources after PCI initialization has occured to handle
- * IO APICS that are mapped in on a BAR in PCI space. */
-late_initcall(ioapic_insert_resources);
 #endif /* !CONFIG_XEN */
--- sle11-2009-10-08.orig/arch/x86/pci/i386.c	2009-10-08 12:08:34.000000000 +0200
+++ sle11-2009-10-08/arch/x86/pci/i386.c	2009-10-08 12:19:13.000000000 +0200
@@ -228,12 +228,14 @@ void __init pcibios_resource_survey(void
 	pcibios_allocate_bus_resources(&pci_root_buses);
 	pcibios_allocate_resources(0);
 	pcibios_allocate_resources(1);
+#ifndef CONFIG_XEN
 	/*
 	 * Insert the IO APIC resources after PCI initialization has
 	 * occured to handle IO APICS that are mapped in on a BAR in
 	 * PCI space, but before trying to assign unassigned pci res.
 	 */
 	ioapic_insert_resources();
+#endif
 }
 
 /**
